avocado/core/test.py

```diff
@@ -133,7 +133,156 @@ def str_filesystem(self):
+            self._cached_test_datadir = os.path.join(file_level_datadir,
+                                                     datadir)
+        return self._cached_test_datadir
+
+    def _get_variant_datadir(self):
+        if self._cached_variant_datadir is None:
+            test_data_dir = self._get_test_datadir()
+            if test_data_dir is None:
+                return None
+            datadir = astring.string_to_safe_path(self.name.variant)
+            self._cached_variant_datadir = os.path.join(test_data_dir,
+                                                        datadir)
+        return self._cached_variant_datadir
+
+    def _get_data_file(self, filename, must_exist):
+        datadir = self._get_file_datadir()
+        if datadir is None:
+            return None
+        file_filename = os.path.join(self.datadir, filename)
+        if not must_exist:
+            self.log.debug(self.GET_DATA_LOG_FMT, filename, file_filename,
+                           "to be found at file-level data dir")
+            return file_filename
+        else:
+            if os.path.exists(file_filename):
+                self.log.debug(self.GET_DATA_LOG_FMT, filename, file_filename,
+                               "found at file-level data dir")
+                return file_filename
+
+    def _get_data_test(self, filename, must_exist):
+        if self.datadir is None or not os.path.isdir(self.datadir):
+            self.log.debug(self.GET_DATA_LOG_FMT, filename, None,
+                           "no data dir exists")
+            return None
+
+        test_data_dir = self._get_test_datadir()
+        test_filename = os.path.join(test_data_dir, filename)
+        if not must_exist:
+            self.log.debug(self.GET_DATA_LOG_FMT, filename, test_filename,
+                           "to be found at test-level data dir")
+            return test_filename
+        else:
+            if os.path.exists(test_filename):
+                self.log.debug(self.GET_DATA_LOG_FMT, filename, test_filename,
+                               "found at test-level data dir")
+                return test_filename
+
+    def _get_data_variant(self, filename, must_exist):
+        if self.name.variant is None:
+            return None
+        variant_data_dir = self._get_variant_datadir()
+        if variant_data_dir is None:
+            return None
+        variant_filename = os.path.join(variant_data_dir, filename)
+        if not must_exist:
+            self.log.debug(self.GET_DATA_LOG_FMT, filename, variant_filename,
+                           "to be found at variant-level data dir")
+            return variant_filename
+        else:
+            if os.path.exists(variant_filename):
+                self.log.debug(self.GET_DATA_LOG_FMT, filename,
+                               variant_filename, "found at variant-level data dir")
+                return variant_filename
+
+    def get_data(self, filename, source=None, must_exist=True):
```

Comment:
I don't really like the idea of `source` being specified from test. Do you see a reason where user would hardcode in their test that they only want to get file from certain location?

As for `must_exist` yes, it's one of the solutions to get the base location (eg. for stdout recording purposes)

Reply:
Just to clarify, I'd like to NACK the `source`. I can live with `sources` but I'd way better prefer not allowing to filter by `source` at least for now. It can always be added later (as I do understand the benefit in having the names available internally for testclass-writing purposes (not for test-writing purposes))